crates/ironrdp-svc/src/lib.rs

```diff
@@ -126,81 +133,132 @@ pub trait StaticVirtualChannel: AsAny + fmt::Debug + Send + Sync {
+    pub fn chunkify(&self, messages: Vec<SvcMessage>, max_chunk_len: usize) -> PduResult<Vec<WriteBuf>> {
+        let mut results = vec![];
+        for message in messages {
+            results.extend(self.chunkify_one(message, max_chunk_len)?);
+        }
+        Ok(results)
+    }
 
-            ChannelPduHeader {
-                length: ironrdp_pdu::cast_int!(ChannelPduHeader::NAME, "length", total_len)?,
-                flags,
-            }
-        };
+    /// Dechunkify a payload received on the virtual channel.
+    ///
+    /// If the payload is not chunked, returns the payload as-is.
+    /// For chunked payloads, returns `Ok(None)` until the last chunk is received, at which point
+    /// it returns `Ok(Some(payload))`.
+    pub fn dechunkify(&mut self, payload: &[u8]) -> PduResult<Option<Vec<u8>>> {
+        // Ensure we have a chunked_pdu buffer
+        if self.chunked_pdu.is_none() {
+            self.chunked_pdu = Some(Vec::new());
+        }
+
+        let last = self.process_header(payload)?;
 
-        // Encode the header for this chunk.
-        encode_buf(&header, &mut chunk)?;
-        // Append the piece of the encoded_pdu that belongs in this chunk.
-        chunk.write_slice(&encoded_pdu[chunk_start_index..chunk_end_index]);
-        // Push the chunk onto the results.
-        chunks.push(chunk);
+        // Extend the chunked_pdu buffer with the payload
+        self.chunked_pdu.as_mut().unwrap().extend_from_slice(payload); // unwrap ok because we just ensured we have a chunked_pdu buffer above
 
-        // If this was the last chunk, we're done, return the results.
+        // If this was an unchunked message, or the last in a series of chunks, return the payload
         if last {
-            break;
+            return Ok(self.chunked_pdu.take());
+        }
+
+        Ok(None)
+    }
+
+    /// Returns whether this was the last chunk based on the flags in the channel header.
+    fn process_header(&self, payload: &[u8]) -> PduResult<bool> {
+        let channel_header = ironrdp_pdu::rdp::vc::ChannelPduHeader::from_buffer(payload)
+            .map_err(|e| custom_err!("failed to decode svc channel header", e))?;
+        Ok(channel_header.flags.contains(ChannelControlFlags::FLAG_LAST))
+    }
+
+    /// Takes a single PDU and breaks it into chunks prefixed with a [`ChannelPduHeader`].
+    ///
+    /// Each chunk is at most `max_chunk_len` bytes long (not including the Channel PDU Header).
+    ///
+    /// For example, if the PDU is 4000 bytes long and `max_chunk_len` is 1600, this function will
+    /// return 3 chunks, each 1600 bytes long, and the last chunk will be 800 bytes long.
+    ///
+    /// [[ Channel PDU Header | 1600 bytes of PDU data ] [ Channel PDU Header | 1600 bytes of PDU data ] [ Channel PDU Header | 800 bytes of PDU data ]]
+    fn chunkify_one(&self, message: SvcMessage, max_chunk_len: usize) -> PduResult<Vec<WriteBuf>> {
```

Comment:
Looks like there is no benefit for `chunkify` and `chunkify_one` to be methods of `SvcProcessor`, the `&self` parameter is not used

Reply:
Technically true, I added these as `&self` methods primarily for having the symmetry of `preprocessor.chunkify` vs `preprocessor.dechunkify` (or `chunk_processor.chunkify`/`chunk_processor.dechunkify` after I rename). I prefer this style but will change it if you _truly_ insist :). For now I've compromised: 4dad3b2486539038cee09cc3d6e7614b6fe2834f

Reply:
Fair enough, yes, I generally tend to prefer removing non required dependencies so we don’t end up with a "Banana Gorilla Jungle" problem and my reflex would be to keep `chunkify` as a free-standing function, unrelated to `SvcPreprocessor`, like it was originally. Also, I see `SvcProcessor` as a "chunk accumulator", so I don’t feel bad if it doesn’t deal with the inverse operation.

That’s my take on it, but I understand your desire to keep the symmetry, and in this case it’s reasonable, so you can consider that I’m not _truly_ insisting :stuck_out_tongue: